### ISSUE TITLE
IDE0005 doesn't find all unused directives

### ISSUE THREAD

**Issue by Evangelink** [dotnet/roslyn#43742]
**Version Used**: VS 16.5.4

I haven't yet been able to create a separate reproducer but on my company's application Resharper or SonarC# are able to find more using directives that can be removed.

**Comment by jmarolf** [comment]
Would be really interested to know the repro steps for this one. The unused directives are issued by the C# compiler itself since it keeps track of what it used for binding types during the compilation

R# has its own implementation of the C# compiler so it may have a different internal definition of what it means for a using to be unused but SonarC# should be using the same apis as the compiler.

**Comment by Evangelink** [comment]
@jmarolf I'll keep you updated as soon as I manage to have a repro. In the meantime the analyzer implementation for SonarC# is https://github.com/SonarSource/sonar-dotnet/blob/master/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnnecessaryUsings.cs

**Comment by Evangelink** [comment]
@jmarolf I can't manage to create a small repro from the code that was causing the issue. I have some knowledge of analyzers so hopefully I will be able to debug the analyzer on my full code and see what's going-on. What's the best-way to create a local debuggable build?

**Comment by jmarolf** [comment]
@Evangelink circling back on this one, this looks like a sonar source bug. Instead of asking the compiler for all diagnostics and looking for instances of `CS8019 Unnecessary using directive` they are attempting to walk the tree itself. See the code [here](https://github.com/dotnet/roslyn/blob/399fa00e26eeffc0d312c03eabafb1c45dbd7521/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Helpers/RemoveUnnecessaryImports/CSharpUnnecessaryImportsProvider.cs#L23) for how the sonar source analyzer should call the compiler.

**Comment by Evangelink** [comment]
@jmarolf I am fine saying the sonar implementation is not ideal/good but when using it, I haven't (yet) been experiencing false-positives AND I have seen it reporting more unused directives than `IDE0005`. I have experienced this behavior a couple of times with an internal project (using WPF), I haven't been able to create a reproducer so I can understand this issue being closed.

**Comment by jmarolf** [comment]
Sometime the compiler will mark a using as "used" because it participates in things like overload resolution. The diagnostic the compiler emits basically means "This using was not involved in binding". There are lots of cases where a using can become involved implicitly and the compiler should correctly list those as "used" using statements.  

It is of course possible that there are bugs in the compiler. If you can provide some kind of reproduceable situation where the compiler labels a using as "necessary" when it is clearly not there is a good chance we would be able to fix it.

**Comment by Evangelink** [comment]
Without bothering you too much, what's the easiest way to debug the analyzer on my company's code, this might actually give me indications on why the using was ignored and so give me an insight to create a repro.

**Comment by jmarolf** [comment]
So on the roslyn side you would want to actually look into the compiler itself.

here is the method that I would want to debug in the compiler:
https://github.com/dotnet/roslyn/blob/23e5b1c1aef5b380c45876a4178110fd5507341d/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L2192

You can start by syncing to the the branch on roslyn that matches whichever VS you have installed. If its VS 16.6 you can do [release/dev16.6-vs-deps](https://github.com/dotnet/roslyn/tree/release/dev16.6-vs-deps).

From there you should be able to set a breakpoint at attach to a running Visual Studio process.

**Comment by Evangelink** [comment]
@jmarolf By "luck" I now have a reproducer for this issue: https://github.com/dotnet/roslyn-analyzers/pull/4247#discussion_r497532471

**Comment by jmarolf** [comment]
@Evangelink can you paste the SHA to the repo here? Thanks!

**Comment by Youssef1313** [comment]
@jmarolf It should be this commit https://github.com/dotnet/roslyn-analyzers/commit/a7211d0a3f3770d3b0aa860b38b4f4c662bb5ebf

**Comment by Evangelink** [comment]
Copy-pasting the comments from @CyrusNajmabadi:

> the copmiler may be seeing a type in that and considering it during lookup while doing lambda overload resolution. If you comment out your context => lambda does this go away?

> if so, this is one of those cases where the namespace is marked by the compiler because it did examien it, but eventually ended up with a resolution taht did not use anything from it. if so, this is a known limitation/extremely-difficult-problem in this space.

**Comment by jmarolf** [comment]
the fact that it's hard to solve doesn't make it not a bug :)

**Comment by Evangelink** [comment]
Commenting lines 37 to 62 doesn't trigger the issue. Playing around with commenting portions of the code I can say that the issue is being triggered when I comment the `catchClause.SemanticModel.AnalyzeXXX` part. To be even more precise accessing `catchClause.SemanticModel` is ok (i.e. issue is raised) but as soon as I access `AnalyzeDataFlow` then the issue disappear.

**Comment by CyrusNajmabadi** [comment]
It's not that it's hard to solve.  It's that, as per language rules, teh compiler does consider the namespace used.

That you can remove the namespace and get the same IL is where the confusion comes from.  Is a namespace used if it doesn't impact IL?  Note that a using may be necessary, even if no types/namespaces are actually used from it.  i.e. if you remove it, even though none of its types/namespaces are used, it will change the meaning of the final code.

**Comment by CyrusNajmabadi** [comment]
> the fact that it's hard to solve doesn't make it not a bug :)

We have 3 choices:

1. IDE computes this.  It doesn't replicate all the compiler rules truly.  We remove usings in some cases where they're needed.  This is bad esp as it can change code meaning.  i.e. the 'using' is needed (even if no types/namespaces are consumed from it) to get teh current IL emitted.
2. Compiler computes this.  It replicates all the rules truly.  But it may consider a using 'seen' even if it could be removed.  This ensure no code semantic changing, but may leave behind a using in complex cases (generally lambda overload resolution).
3. Compiler computes this, and truly figure out if removing the using would lead to any IL change.  This is too costly to determine with 100% accuracy given the design of hte compiler.

**Comment by jmarolf** [comment]
@CyrusNajmabadi I agree that the compiler team should investigate first and decide if everything is working as expected. For 2. I do not yet understand why the compiler can't implement a refinement that if a using is used only for overload resolution but not selected that it is not marked as "seen." still I will defer to the compiler team here.

**Comment by CyrusNajmabadi** [comment]
> I do not yet understand why the compiler can't implement a refinement that if a using is used only for overload resolution but not selected that it is not marked as "seen."

Because removal *might* change the meaning, and the compiler doesn't really have any mechanism to determine that cheaply.  It's complex to tell, but imagine the following:

```
using System;

using X;
using Y;

class B
{
    static void Main()
    {
        Bar(x => x.Goo(), null); // Prints 1
    }

    static void Bar(Action<string> x, object y)
    {
        Console.WriteLine(1);
    }

    static void Bar(Action<int> x, string y)
    {
        Console.WriteLine(2);
    }
}

namespace X
{
    public static class A
    {
        public static void Goo(this int x)
        {
        }

        public static void Goo(this string x)
        {
        }
    }
}

namespace Y
{
    public static class B
    {
        public static void Goo(this int x)
        {
        }
    }
}
```

Here, nothing from namespace `Y` is actually used.  i.e. the call to Goo in `Bar(x => x.Goo(), null);` is a call to X.A.Goo.  *However* if you remove the `using Y;` then this code changes meaning and the second 'Bar' is called rather than the first.  The reason for this is that a `using` can do more than bring in types/namespaces/extensions.  Importantly, a using can bring in 'conflicts'.  Those conflicts end up causing the compiler to reject teh second overload and pick the first.  If you remove the using, you remove the conflict and hte second overload becomes better.

**Comment by Evangelink** [comment]
Feel free to close the issue as "Won't Fix".
